gnucash/import-export/bi-import/dialog-bi-import.c

```diff
@@ -263,126 +268,100 @@ gnc_bi_import_fix_bis (GtkListStore * store, guint * fixed, guint * deleted,
+                {
+                    if (!gnc_search_vendor_on_id
+                        (gnc_get_current_book (), owner_id))
+                    {
+                        // Vendor not found.
+                        valid = gtk_list_store_remove (store, &iter);
+                        row_deleted = TRUE;
+                        g_string_append_printf (info,
+                                                _("ROW %d DELETED, VENDOR_DOES_NOT_EXIST: id=%s\n"),
+                                                row, id);
+                    }
+                }
+                else if (g_ascii_strcasecmp (type, "INVOICE") == 0)
+                {
+                    if (!gnc_search_customer_on_id
+                        (gnc_get_current_book (), owner_id))
+                    {
+                        // Customer not found.
+                        valid = gtk_list_store_remove (store, &iter);
+                        row_deleted = TRUE;
+                        g_string_append_printf (info,
+                                                _("ROW %d DELETED, CUSTOMER_DOES_NOT_EXIST: id=%s\n"),
+                                                row, id);
+                    }
+                }
+            }
+            
+            // Verify the date opened.
+            if(!row_deleted && !isDateValid(date_opened))
+            {
+                // Fix this by using the current date.
                 gchar temp[20];
                 GDate date;
                 g_date_clear (&date, 1);
                 gnc_gdate_set_today (&date);
-                g_date_strftime (temp, 20, date_format_string, &date);	// Create a user specified date string.
+                g_date_strftime (temp, 20, date_format_string, &date);    // Create a user specified date string.
                 gtk_list_store_set (store, &iter, DATE_OPENED,
                                     temp, -1);
                 row_fixed = TRUE;
             }
-            
-            // date_opened is valid
-
-            /*
-            // Check if due date is valid.  Set it to date_posted if not valid or missing.
-            if(!isDateValid(due_date))
-            {
-                gtk_list_store_set (store, &iter, DUE_DATE,
-                                        date_posted, -1);
-                row_fixed = TRUE;
-
-            }
-
-            // due_date is valid
-            */
-            if (strlen (quantity) == 0)
-            {
-                // quantity is unset => set to 1
-                gtk_list_store_set (store, &iter, QUANTITY, "1", -1);
-                row_fixed = TRUE;
-            }
-
-
-            // quantity is valid
 
-            if (strlen (owner_id) == 0)
+        }
+        
+        // Validate and fix item data.
+        if (!row_deleted)
+        {
+            // Validate the price.
+            if (strlen (price) == 0)
             {
+                // No valid price, delete the row
```

Comment:
This is a tricky thing to do: you're validating the first line of a potentially multi-line invoice definition. All the invoice's header fields validated fine in this case. It's only the entry fields on this line that don't validate.
If you now remove the line based on that, you risk that all subsequent lines will be removed as well as they may not have invoice header data at all.
In contrast if a non-first line is missing a price, only that line will be deleted from the invoice. This results in an inconsistent user experience.

I actually think the whole concept of simply removing lines automatically give a poor user experience. It would be much better if the user is shown the bad lines with an explanation of why they were considered bad. That would greatly help the user go in and fix the csv file. But that may be a future improvement. It's not blocking this PR for me.

Reply:
Yes, I agree that this results in an inconsistent user experience. And I could have added code that takes the validated header data of a first row, and uses it for subsequent rows of the same invoice, even if this first row is deleted because the item data does not validate.
However, in my view, if an invoice has one row that does not validate, you should reject the complete invoice; else you will import only the correct items of the invoice, which leads to an incorrect total amount of that invoice in the system. I implemented that functionality (all invoice items should validate, or the invoice will not be imported) in the next commit. So I did not put the effort into solving the somewhat inconsistent user experience here, because I implemented a different and consistent user experience in the next commit.
I also agree that it would be better to show all bad lines to the user, but, as you say, that may be a future improvement. For now, I at least improved the user feedback on the first erroneous row, in a subsequent commit of the pull request.